# Individual Playwrights in the Context of The Evolution of The French Comedy
In this analysis we will explore the playwrights in the context of the evolution of the five-act comedy in verse. We will focus on the following features:
- The number of dramatic characters
- The mobility coefficient
- The standard range of the number of speaking characters
- The percentage of polylogues
- The percentage of monologues.

Our agenda:
1. For each time period and each feature, we will identify the comedy that that represents the maximum, the minimum, and the comedies that are the closest to the mean.
2. For each time period, we create a summary for individual playwrights that represent two groups (more prolific, i.e., those who wrote at least two five-act verse comedies, and less prolific, i.e., those who wrote only one such comedy).
    - Calculate the mean number of dramatic characters, mean mobility coefficient, the standard range of the number of speaking characters (sigma), the mean percentage of polylogues, and the mean percentage of monologues.
    - Calculate open-form scores for each playwright that tell us how experimental each playwright is.
3. Examine the evolution of Jean-François Collin d’Harleville whose oeuvre spanned two literary periods.

### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import re

### Define Functions

In [2]:
# update the boundaries as we determined based on our hypothesis testing
def determine_period(row):
    if row <= 1695:
        period = 1
    elif row >= 1696 and row <= 1795:
        period = 2
    else:
        period = 3
    
    return period

In [3]:
def make_list(row):
    speech_dist = []
    for value in row[1:-1].split('\n '):
        speech_dist.append([int(num) for num in re.findall('[0-9]+', value)])
        
    return speech_dist

In [4]:
def speech_distribution_by_period(period_df):
    all_distributions = []
    for row in period_df['speech_distribution']:
        speech_dist_df = pd.DataFrame(row).T
        # rename columns to make sure they start with 1 and not 0
        speech_dist_df.columns = speech_dist_df.iloc[0, :]
        # no need to include the variants as a row - they will be column names
        only_counts_df = pd.DataFrame(speech_dist_df.iloc[1, :])
        only_counts_df.columns = ['raw_numbers']
        only_counts_df['percentage'] = only_counts_df['raw_numbers'] / only_counts_df.sum().values[0]
        all_distributions.append(round(only_counts_df['percentage'], 4))
    period_df_dist = pd.concat(all_distributions, axis=1).fillna(0)
    # take the mean for each period
    mean_per_type = pd.DataFrame(period_df_dist.mean(axis=1)).T 
    mean_per_type.index.name = 'number_of_speakers'
    mean_per_type = (mean_per_type * 100).round(2)
        
    return mean_per_type

In [5]:
def sigma_iarkho(df):
    """
    The function allows calculating standard range following iarkho's procedure.
    Parameters:
        df  - a dataframe where columns are variants, i.e., the distinct number of speakers in the ascending order, 
              e.g. [1, 2, 3, 4, 5] and values weights corresponding to these variants, i.e.,
              the number of scenes, e.g. [20, 32, 18, 9, 1]
    Returns:
        sigma - standard range per iarkho
    """
    weighted_mean_variants = np.average(df.columns.tolist(), weights=df.values[0])
    differences_squared = [(variant - weighted_mean_variants)**2 for variant in df.columns]
    weighted_mean_difference = np.average(differences_squared, weights=df.values[0])
    sigma = round(weighted_mean_difference**0.5, 2)

    return sigma

In [6]:
def sigma_for_playwrights(df, playwrights_lst):
    sigmas = []
    for playwright in playwrights_lst:
        selection = df[(df.last_name == playwright[0]) & (df.first_name == playwright[1])].copy()
        sigma = selection.pipe(speech_distribution_by_period).pipe(sigma_iarkho)
        sigmas.append(sigma)
        
    summary = pd.DataFrame(sigmas, columns=['sigma_iarkho'])
    summary['z_score'] = (summary['sigma_iarkho'] - df['sigma_iarkho'].mean()) / df['sigma_iarkho'].std()
    summary.index = playwrights_lst
    
    return summary

In [7]:
def summary_feature(df, feature):
    print('Mean, standard deviation, median, min and max values for the period:')
    display(df[feature].describe()[['mean', 'std', '50%','min', 'max']].round(2))
    print('Period Max:')
    display(df[df[feature] == df[feature].max()][['last_name', 'first_name', 'title', 'date', feature]].round(2))
    print('Period Min:')
    display(df[df[feature] == df[feature].min()][['last_name', 'first_name', 'title', 'date', feature]].round(2))
    print('The closest to the mean:')
    df_copy = df.copy()
    df_copy['diff_with_mean'] = df_copy[feature].apply(lambda x: np.absolute(x - df_copy[feature].mean()))
    display(df_copy[df_copy['diff_with_mean'] == df_copy['diff_with_mean'].min()][['last_name', 'first_name', 'title', 'date', feature]].round(2))

In [8]:
def authors_data(data_df, feature):
    overall_mean = round(data_df[feature].mean(), 2)
    overall_std = round(data_df[feature].std(), 2)
    statistics = ['mean'] 
    all_authors = pd.DataFrame(data_df.groupby(['last_name', 'first_name'])[feature].mean())
    all_authors.columns= ['mean']
    all_authors['z_score'] = (all_authors['mean'] - overall_mean) / overall_std
    
    return  all_authors

In [9]:
def playwrights_place(df, with_z_score=True):
    if with_z_score:
        column = 'z_score'
        sigma_col = column
    else:
        column = ['mean']
        sigma_col = 'sigma_iarkho'
    summary = pd.DataFrame(authors_data(df, 'num_present_characters')[column])
    summary.columns = ['num_present_characters']
    # make sure the order of the playwrights is the same
    
    ind = summary.index
    summary['mobility_coefficient'] = authors_data(df, 'mobility_coefficient', 
                                                        ).loc[ind, column]
    summary['sigma_iarkho'] = sigma_for_playwrights(df, ind)[sigma_col]
    summary['polylogues'] = authors_data(df, 'percentage_polylogues', 
                                                         ).loc[ind, column]
    summary['monologues'] = authors_data(df, 'percentage_monologues', 
                                                         ).loc[ind, column]
    summary = summary.round(2)
    if with_z_score:
        summary['monologues'] = summary['monologues'].apply(lambda x: -x)
        summary['open_form_score'] = round(summary.apply(lambda x: x.mean(), axis=1), 2)
        summary = summary.sort_values(by='open_form_score', ascending=False)
        
    return summary

### Load Data

In [10]:
data = pd.read_csv('../French_Comedies/Data/French_Comedies_Data.csv')

In [11]:
data.shape

(279, 25)

In [12]:
# include only five act comedies and only original comedies
original_comedies = data[(data['num_acts'] ==5)&
                         (data['translation/adaptation/contrastive'] == 0)].copy()

In [13]:
original_comedies.head(3)

index                                 title  last_name first_name  date  \
2   F_3         Mélite ou Les fausses lettres  Corneille     Pierre  1629   
3   F_5  La Veuve ou Le Traître trahi          Corneille     Pierre  1633   
4   F_9                           La Célimène     Rotrou    Jean de  1633   

   translation/adaptation/contrastive  num_acts  \
2                                   0         5   
3                                   0         5   
4                                   0         5   

                                                 url  num_present_characters  \
2  http://www.theatre-classique.fr/pages/document...                       8   
3  http://www.theatre-classique.fr/pages/document...                      12   
4  http://www.xn--thtre-documentation-cvb0m.com/c...                      10   

   num_scenes_text  ...  percentage_above_two_speakers  \
2               35  ...                          23.08   
3               40  ...                          20.00   
4               36  ...                          22.22   

  av_percentage_non_speakers  sigma_iarkho  \
2                      0.513         0.906   
3                      3.519         1.062   
4                      8.963         1.092   

   number_scenes_with_discontinuous_change_characters  \
2                                                 12    
3                                                 12    
4                                                  5    

   percentage_scenes_with_discontinuous_change_characters  total_utterances  \
2                                             30.769                  483.0   
3                                             26.667                  521.0   
4                                             11.111                    NaN   

   num_verse_lines  dialogue_vivacity  five_year_intervals  decades  
2           1822.0              0.265                 1630     1630  
3           2010.0              0.259                 1635     1640  
4              NaN                NaN                 1635     1640  

[3 rows x 25 columns]

In [14]:
original_comedies.shape

(257, 25)

In [15]:
sorted_comedies = original_comedies.sort_values(by='date')

In [16]:
sorted_comedies['period'] = sorted_comedies['date'].apply(determine_period)

In [17]:
sorted_comedies = sorted_comedies.rename(columns={'num_scenes_iarkho': 'mobility_coefficient', 
                                                  'percentage_non_duologues': 'percentage_non_dialogues',
                                                  'percentage_above_two_speakers': 'percentage_polylogues'})

In [18]:
# remove white spaces
sorted_comedies['last_name'] = sorted_comedies['last_name'].str.strip()
sorted_comedies['first_name'] = sorted_comedies['first_name'].str.strip()
sorted_comedies['first_name'] = sorted_comedies['first_name'].fillna('')
sorted_comedies['speech_distribution'] = sorted_comedies['speech_distribution'].apply(make_list)

In [19]:
period_one = sorted_comedies[sorted_comedies.period == 1].copy()
period_two = sorted_comedies[sorted_comedies.period == 2].copy()
period_three = sorted_comedies[sorted_comedies.period == 3].copy()

## Part 1. The Maximum, The Minimum, and The Closest to The Mean

### Period 1 (1629 to 1695)

#### Number of Dramatic Characters

In [20]:
summary_feature(period_one, 'num_present_characters')

Mean, standard deviation, median, min and max values for the period:


mean    11.91
std      4.05
50%     11.00
min      7.00
max     30.00
Name: num_present_characters, dtype: float64

Period Max:


last_name       first_name                title  date  \
80  Chalussay  Le Boulanger de  Elomire Hypocondre   1670   

    num_present_characters  
80                      30

Period Min:


last_name first_name                    title  date  \
25   Corneille     Pierre      La Suite du menteur  1643   
70  Montfleury                  L’École des filles  1666   
81      Marcel             Le Mariage sans mariage  1671   

    num_present_characters  
25                       7  
70                       7  
81                       7

The closest to the mean:


last_name            first_name                                 title  \
3    Corneille                Pierre  La Veuve ou Le Traître trahi           
6    Corneille                Pierre                  La Galerie du Palais   
54  Boisrobert  François Le Métel de             Les Apparences trompeuses   
60     Molière                                           Le Dépit amoureux   
65  Montfleury                                          Le Mari sans femme   
66   Boursault                  Edmé                         Les Nicandres   
68   Boursault                  Edmé               Les deux frères gémeaux   
79     Poisson               Raymond  Les Pipeurs ou les Femmes coquettes    

    date  num_present_characters  
3   1633                      12  
6   1634                      12  
54  1655                      12  
60  1658                      12  
65  1663                      12  
66  1663                      12  
68  1665                      12  
79  1670                      12

#### Mobility Coefficient

In [21]:
summary_feature(period_one, 'mobility_coefficient')

Mean, standard deviation, median, min and max values for the period:


mean    41.9
std     11.6
50%     40.0
min     19.0
max     85.0
Name: mobility_coefficient, dtype: float64

Period Max:


last_name               first_name             title  date  \
86  Hauteroche  Noël Lebreton, sieur de  Crispin musicien  1674   

    mobility_coefficient  
86                    85

Period Min:


last_name first_name                      title  date  \
18  Guérin de Bouscal      Guyon  Dom Quixote de la Manche   1639   

    mobility_coefficient  
18                    19

The closest to the mean:


last_name            first_name                                 title  \
39  Boisrobert  François Le Métel de                      La Folle Gageure   
45   Corneille                Thomas                  Le Charme de la voix   
79     Poisson               Raymond  Les Pipeurs ou les Femmes coquettes    

    date  mobility_coefficient  
39  1651                    42  
45  1653                    42  
79  1670                    42

#### Standard Range of the Number of Speaking Characters (Sigma)

In [22]:
summary_feature(period_one, 'sigma_iarkho')

Mean, standard deviation, median, min and max values for the period:


mean    1.25
std     0.40
50%     1.19
min     0.60
max     3.54
Name: sigma_iarkho, dtype: float64

Period Max:


last_name       first_name                title  date  sigma_iarkho
80  Chalussay  Le Boulanger de  Elomire Hypocondre   1670          3.54

Period Min:


last_name first_name             title  date  sigma_iarkho
69  Quinault   Philippe  La Mère coquette  1665           0.6

The closest to the mean:


last_name first_name                title  date  sigma_iarkho
89  Montfleury             Crispin gentilhomme  1677          1.25

#### The Percentage of Polylogues

In [23]:
summary_feature(period_one, 'percentage_polylogues')

Mean, standard deviation, median, min and max values for the period:


mean    42.36
std     13.87
50%     41.46
min     10.64
max     86.21
Name: percentage_polylogues, dtype: float64

Period Max:


last_name            first_name              title  date  \
41  Boisrobert  François Le Métel de  Les trois Orontes  1652   

    percentage_polylogues  
41                  86.21

Period Min:


last_name first_name        title  date  percentage_polylogues
7  Corneille     Pierre  La Suivante  1634                  10.64

The closest to the mean:


last_name  first_name             title  date  percentage_polylogues
26   Ouville  Antoine d'  La Dame suivante  1643                  42.31

#### The Percentage of Monologues

In [24]:
summary_feature(period_one, 'percentage_monologues')

Mean, standard deviation, median, min and max values for the period:


mean    13.41
std      9.55
50%     12.12
min      0.00
max     33.33
Name: percentage_monologues, dtype: float64

Period Max:


last_name first_name        title  date  percentage_monologues
4    Rotrou    Jean de  La Célimène  1633                  33.33

Period Min:


last_name            first_name  \
18  Guérin de Bouscal                 Guyon   
35            Scarron                  Paul   
42            Scarron                  Paul   
46         Boisrobert  François Le Métel de   
61          Corneille                Thomas   
69           Quinault              Philippe   
71            Molière                         
73          Corneille                Thomas   
77          Corneille                Thomas   

                                        title  date  percentage_monologues  
18                  Dom Quixote de la Manche   1639                    0.0  
35  L'Héritier ridicule ou la Dame intéressée  1649                    0.0  
42                       Don Japhet d'Arménie  1652                    0.0  
46                         La Belle plaideuse  1654                    0.0  
61                           Le Galant doublé  1660                    0.0  
69                           La Mère coquette  1665                    0.0  
71                             Le Misanthrope  1666                    0.0  
73                        Le Baron d'Albikrac  1668                    0.0  
77                      La Comtesse d'Orgueil  1670                    0.0

The closest to the mean:


last_name first_name                          title  date  \
19  Mareschal      André  Le Véritable Capitan Matamore  1640   

    percentage_monologues  
19                  13.33

## Period 2 (1696 to 1795)

#### Number of Dramatic Characters

In [25]:
summary_feature(period_two, 'num_present_characters')

Mean, standard deviation, median, min and max values for the period:


mean    10.53
std      2.89
50%     10.00
min      6.00
max     24.00
Name: num_present_characters, dtype: float64

Period Max:


last_name first_name                      title  date  \
201  Cubières-Palmézeaux  Michel de  L'Homme d'état imaginaire  1789   

     num_present_characters  
201                      24

Period Min:


last_name        first_name              title  date  \
158  Néricault Destouches                    Le Mari confident  1758   
163               Bastide  Jean-François de     Le Jeune homme  1766   

     num_present_characters  
158                       6  
163                       6

The closest to the mean:


last_name     first_name  \
106              Dancourt                  
126  Néricault Destouches                  
128             Du Fresny        Charles   
132              Rousseau  Jean-Baptiste   
138  Néricault Destouches                  
187                Bièvre     Marquis de   
192                 Borel                  
197   Collin d’Harleville  Jean-François   
202     Fabre d'Églantine                  

                                                 title  date  \
106       Les Enfants de Paris ou la Famille à la mode  1699   
126                           Les Philosophes amoureux  1729   
128                                    Le Faux Sincère  1731   
132  Les ayeux chimériques ou la comtesse de Critognac  1735   
138                        L'Ambitieux et l'Indiscrète  1737   
187                                       Le Séducteur  1783   
192                                         Le Méfiant  1786   
197            L'Optimiste ou l'Homme content de tout   1788   
202                             Le Philinte de Molière  1790   

     num_present_characters  
106                      11  
126                      11  
128                      11  
132                      11  
138                      11  
187                      11  
192                      11  
197                      11  
202                      11

#### Mobility Coefficient

In [26]:
summary_feature(period_two, 'mobility_coefficient')

Mean, standard deviation, median, min and max values for the period:


mean    49.29
std     11.32
50%     48.50
min     29.00
max     91.00
Name: mobility_coefficient, dtype: float64

Period Max:


last_name first_name            title  date  mobility_coefficient
164  Chauveau             L'homme de cour  1767                    91

Period Min:


last_name     first_name     title  date  \
142  Nivelle de la Chaussée  Pierre Claude  Mélanide  1741   

     mobility_coefficient  
142                    29

The closest to the mean:


last_name     first_name  \
104              Rousseau  Jean-Baptiste   
113               Regnard  Jean-François   
124  Néricault Destouches                  
140                Boissy       Louis de   
148              Voltaire                  
156               La Noue  Jean-Baptiste   
173                  Bret        Antoine   
209               Reynier             L.   

                                              title  date  \
104                                     Le Flatteur  1696   
113                          Le Légataire Universel  1708   
124                             Le Philosophe marié  1727   
140                            Les Dehors trompeurs  1740   
148                                        La Prude  1747   
156                            La Coquette corrigée  1757   
173  Le protecteur bourgeois ou la confiance trahie  1772   
209                                L'Avare fastueux  1794   

     mobility_coefficient  
104                    49  
113                    49  
124                    49  
140                    49  
148                    49  
156                    49  
173                    49  
209                    49

#### Sigma

In [27]:
summary_feature(period_two, 'sigma_iarkho')

Mean, standard deviation, median, min and max values for the period:


mean    1.12
std     0.29
50%     1.10
min     0.63
max     1.98
Name: sigma_iarkho, dtype: float64

Period Max:


last_name     first_name                                         title  \
154  Rousseau  Jean-Baptiste  La Femme qui ne parle point ou l'hypocondre    

     date  sigma_iarkho  
154  1751          1.98

Period Min:


last_name     first_name                   title  date  sigma_iarkho
179     Dorat  Claude-Joseph  Roséïde ou l'Intrigant  1779          0.63

The closest to the mean:


last_name first_name        title  date  sigma_iarkho
120  Néricault Destouches             Le Médisant  1715          1.12

#### The Percentage of Polylogues

In [28]:
summary_feature(period_two, 'percentage_polylogues')

Mean, standard deviation, median, min and max values for the period:


mean    31.62
std     11.03
50%     31.50
min      8.70
max     58.06
Name: percentage_polylogues, dtype: float64

Period Max:


last_name     first_name  \
132  Rousseau  Jean-Baptiste   

                                                 title  date  \
132  Les ayeux chimériques ou la comtesse de Critognac  1735   

     percentage_polylogues  
132                  58.06

Period Min:


last_name first_name                                           title  \
180     Delon             Le Financier, comédie en cinq actes et en vers   

     date  percentage_polylogues  
180  1779                    8.7

The closest to the mean:


last_name     first_name               title  date  percentage_polylogues
107   Regnard  Jean-François  Démocrite amoureux  1700                  31.58

#### The Percentage of Monologues

In [29]:
summary_feature(period_two, 'percentage_monologues')

Mean, standard deviation, median, min and max values for the period:


mean    22.08
std      7.75
50%     21.88
min      3.23
max     40.48
Name: percentage_monologues, dtype: float64

Period Max:


last_name        first_name           title  date  percentage_monologues
163   Bastide  Jean-François de  Le Jeune homme  1766                  40.48

Period Min:


last_name     first_name  \
132  Rousseau  Jean-Baptiste   

                                                 title  date  \
132  Les ayeux chimériques ou la comtesse de Critognac  1735   

     percentage_monologues  
132                   3.23

The closest to the mean:


last_name     first_name                  title  date  \
207  Collin d’Harleville  Jean-François  Le Vieux célibataire   1792   

     percentage_monologues  
207                  22.03

### Period Three (1796 to 1847)

#### Number of Dramatic Characters

In [30]:
summary_feature(period_three, 'num_present_characters')

Mean, standard deviation, median, min and max values for the period:


mean    12.11
std      4.28
50%     11.00
min      6.00
max     28.00
Name: num_present_characters, dtype: float64

Period Max:


last_name first_name                                   title  date  \
259     Gosse    Étienne  Les Jésuites, ou les autres Tartuffes   1827   

     num_present_characters  
259                      28

Period Min:


last_name first_name      title  date  num_present_characters
278    Augier      Émile  Gabrielle  1849                       6

The closest to the mean:


last_name         first_name                               title  \
215  Fabre d'Églantine                                        Les Précepteurs   
228            Étienne  Charles-Guillaume                    Les Deux gendres   
242           Théaulon           Emmanuel  L'Artiste ambitieux, ou l'Adoption   
250            Bonjour            Casimir     L’Éducation ou les deux Cousine   
258            Bonjour            Casimir      L’Argent ou les Moeurs du Jour   

     date  num_present_characters  
215  1799                      12  
228  1810                      12  
242  1820                      12  
250  1823                      12  
258  1826                      12

#### Mobility Coefficient

In [31]:
summary_feature(period_three, 'mobility_coefficient')

Mean, standard deviation, median, min and max values for the period:


mean    54.58
std      9.99
50%     53.50
min     34.00
max     86.00
Name: mobility_coefficient, dtype: float64

Period Max:


last_name first_name                                   title  date  \
259     Gosse    Étienne  Les Jésuites, ou les autres Tartuffes   1827   

     mobility_coefficient  
259                    86

Period Min:


last_name first_name                     title  date  mobility_coefficient
240     Duval  Alexandre  Le Chevalier d'industrie  1818                    34

The closest to the mean:


last_name    first_name                             title  date  \
227     Picard  Louis-Benoît  Les Capitulations de conscience   1809   
245  Delavigne       Casimir                     Les Comédiens  1820   

     mobility_coefficient  
227                    55  
245                    55

#### Sigma

In [32]:
summary_feature(period_three, 'sigma_iarkho')

Mean, standard deviation, median, min and max values for the period:


mean    1.33
std     0.32
50%     1.26
min     0.84
max     2.54
Name: sigma_iarkho, dtype: float64

Period Max:


last_name first_name                 title  date  sigma_iarkho
262  Delavigne    Casimir  La Princesse Aurélie  1828          2.54

Period Min:


last_name first_name                                           title  \
241   Michaud      L. G.  Le Faux ami de cour, ou le Danger des liaisons   

     date  sigma_iarkho  
241  1818          0.84

The closest to the mean:


last_name first_name                         title  date  sigma_iarkho
275       Rey    Charles  La séduction et l'amour vrai  1847          1.33

#### The Percentage of Polylogues

In [33]:
summary_feature(period_three, 'percentage_polylogues')

Mean, standard deviation, median, min and max values for the period:


mean    38.09
std      8.70
50%     37.31
min     18.46
max     65.00
Name: percentage_polylogues, dtype: float64

Period Max:


last_name first_name                                              title  \
239  Merville             La Famille Glinet, ou Les premiers temps de la...   

     date  percentage_polylogues  
239  1818                   65.0

Period Min:


last_name first_name                                   title  date  \
252   Bonjour    Casimir  Le Mari à bonnes fortunes, ou La Leçon  1824   

     percentage_polylogues  
252                  18.46

The closest to the mean:


last_name first_name                           title  date  \
258   Bonjour    Casimir  L’Argent ou les Moeurs du Jour  1826   

     percentage_polylogues  
258                  37.88

#### The Percentage of Monologues

In [34]:
summary_feature(period_three, 'percentage_monologues')

Mean, standard deviation, median, min and max values for the period:


mean    20.08
std      5.95
50%     18.94
min     11.36
max     43.08
Name: percentage_monologues, dtype: float64

Period Max:


last_name first_name                                   title  date  \
252   Bonjour    Casimir  Le Mari à bonnes fortunes, ou La Leçon  1824   

     percentage_monologues  
252                  43.08

Period Min:


last_name first_name      title  date  percentage_monologues
278    Augier      Émile  Gabrielle  1849                  11.36

The closest to the mean:


last_name first_name        title  date  percentage_monologues
244     Gosse    Étienne  Le Flatteur  1820                   20.0

### Summary

The maximum for all features:
- Number of Dramatic Characters (period one: 30 (*Elomire Hypocondre* by Chalussay), period two: 24 (*L'Homme d'état imaginaire* by Cubières-Palmézeaux, period three: 	28 (*Les Jésuites, ou les autres Tartuffes*) by Gosse)
    - Mobility Coefficient (period one: 85 (*Crispin musicien* by Hauteroche), period two: 91 (*L'homme de cour*) by Chauveau, period three: 86 (*Les Jésuites, ou les autres Tartuffes* by Gosse))
    - Sigma (period one: 3.54 (*Elomire Hypocondre* by Chalussay), period two: 1.98 (*La Femme qui ne parle point ou l'hypocondre* by Rousseau), period three: 2.54 (*La Princesse Aurélie* by Delavigne))
    - The percentage of polylogues (period one: 86.21% (*Les trois Orontes* by Boisrobert), period two: 58.06% (*Les ayeux chimériques ou la comtesse de Critognac* by Rousseau), period three: 65.0% *La Famille Glinet* by Merville)
    - The percentage of monologues (period one: 33.33% (*La Célimène* by Rotrou), period two: 40.48% (*Le Jeune homme* by Bastide), period three: 43.08% (*Le Mari à bonnes fortunes, ou La Leçon* by Bonjour))
    
   Three authors (Chalussay, Rousseau, and Gosse) wrote the comedies that represent the maximum for their period based on at least two features.
   
The minimum for all features:   
   - Number of Dramatic Characters (period one: 7 (three comedies *La Suite du menteur* by Pierre Corneille, *	L’École des filles* by Montfleury, and *Le Mariage sans mariage* by Marcel), period two: 6 (*Le Mari confident* by Destouches and *Le Jeune homme* by Bastide), period three: 6 (*Gabrielle* by Augier))
    - Mobility Coefficient (period one: 19 (*Dom Quixote de la Manche* by Guérin de Bouscal), period two: 29 (*Mélanide* by Nivelle de la Chaussée), period three: *Le Chevalier d'industrie* by Duval)
    - Sigma (period one: 0.6 (*La Mère coquette* by Quinault), period two: 0.63 (*Roséïde ou l'Intrigant* by Dorat), period three: 0.84 (*Le Faux ami de cour, ou le Danger des liaisons* by Michaud))
    - The percentage of polylogues (10.64% (*La Suivante* by Pierre Corneille), period two: 8.7% (*Le Financier* by Delon), period three: 18.46% (*Le Mari à bonnes fortunes, ou La Leçon* by Bonjour))
    - The percentage of monologues (period one: period one: 0% (nine comedies), period two: 3.23% (*Les ayeux chimériques ou la comtesse de Critognac* by Rousseau) , period three: 11.36% (*Gabrielle* by Augier))
    
    
    Two authors (Pierre Corneille and Casimir Bonjour) wrote comedies that represent the period three minimum based on at least two features.

## Part 2. Summary For Individual Playwrights

We calculate the mean number of dramatic characters, mean mobility coefficient, the standard range of the number of speaking characters (sigma), the mean percentage of polylogues, and the mean percentage of monologues. We quanitify how much a playwright prefers to write comedies with more open forms based on all features.

#### Open Form Scores:
1. For all features, we will calculate **the z-score**: $z=(x-u)/s$ where where u is the mean of the feature and s is the standard deviation of this feature. For the percentage of monologues, we will reverse the sign, i.e., will use **- z-score** since it is the lower value of the percentage of monologues that indicates a more open form.
3. The **open form score** will be the mean z-score. For example, if Boisrobert has the following z-scores (0.07, -0.35, 0.41, 0.68) and -z-score for the percentage of monologues of 0.97, his **open form score** = (0.07 -0.35 +  0.41 + 0.68 + 0.97) / 5 ≈ 0.36. The open form scores can be positive and negative, where high positive number indicates the most open form, whereas the high negative number indicates the least open form.

In [35]:
results = playwrights_place(sorted_comedies, with_z_score=True)

## More and Less Prolific Playwrights

We will define the playwrights who wrote at least two five-act comedies in verse as prolific and those who wrote a single five-act verse comedy as less prolific. We will explore how many representatives are in each group and examine if open-form scores differ in those two groups by period.

In [36]:
def authors_by_output(df):
    counts = df.groupby(['last_name', 'first_name']).count()['index']
    playwrights_over_one = counts[counts > 1].index
    playwrights_one = counts[counts == 1].index
    
    return playwrights_over_one, playwrights_one

In [37]:
over_one, only_one = authors_by_output(sorted_comedies)

In [38]:
sorted_above_one = results.loc[over_one, :].sort_values(by='open_form_score', ascending=False)

#### Prolific

In [39]:
print('Number of more prolific playwrigthts:', 
sorted_above_one.shape[0])

Number of more prolific playwrigthts: 41


In [40]:
print('Number of more prolific playwrigthts writing comedies with the open form:', 
sorted_above_one[sorted_above_one.open_form_score >= 0].shape[0])

Number of more prolific playwrigthts writing comedies with the open form: 30


In [41]:
print('Number of more prolific playwrigthts writing comedies with the closed form:', 
sorted_above_one[sorted_above_one.open_form_score < 0].shape[0])

Number of more prolific playwrigthts writing comedies with the closed form: 11


#### Less Prolific

In [42]:
sorted_only_one = results.loc[only_one, :].sort_values(by='open_form_score', ascending=False)

In [43]:
print('Number of less prolific playwrigthts:', 
sorted_only_one.shape[0])

Number of less prolific playwrigthts: 82


In [44]:
print('Number of less prolific playwrigthts writing comedies with the open form:', 
sorted_only_one[sorted_only_one.open_form_score >= 0].shape[0])

Number of less prolific playwrigthts writing comedies with the open form: 32


In [45]:
print('Number of less prolific playwrigthts writing comedies with the closed form:', 
sorted_only_one[sorted_only_one.open_form_score < 0].shape[0])

Number of less prolific playwrigthts writing comedies with the closed form: 50


## Period One: Z-Scores and Open-Form Scores

### Prolific

In [46]:
sorted_above_one_play_period_one = sorted_above_one.loc[period_one.last_name.unique(), :]

In [47]:
sorted_above_one_play_period_one

num_present_characters  \
last_name                first_name                                        
Hauteroche               Noël Lebreton, sieur de                    0.62   
Boisrobert               François Le Métel de                       0.47   
La Fontaine              Jean de                                    2.02   
Gillet de la Tessonnerie N.                                         0.60   
Ouville                  Antoine d'                                -0.29   
Corneille                Thomas                                    -0.04   
Scarron                  Paul                                       0.02   
Boursault                Edmé                                       1.22   
Molière                                                            -0.07   
Quinault                 Philippe                                  -0.82   
Mareschal                André                                     -0.78   
Baron                                                               0.42   
Montfleury                                                         -0.54   
Corneille                Pierre                                    -0.34   

                                                  mobility_coefficient  \
last_name                first_name                                      
Hauteroche               Noël Lebreton, sieur de                  1.51   
Boisrobert               François Le Métel de                    -0.86   
La Fontaine              Jean de                                  0.09   
Gillet de la Tessonnerie N.                                      -1.25   
Ouville                  Antoine d'                               0.31   
Corneille                Thomas                                  -0.48   
Scarron                  Paul                                    -0.69   
Boursault                Edmé                                    -0.25   
Molière                                                          -1.26   
Quinault                 Philippe                                -0.81   
Mareschal                André                                   -1.31   
Baron                                                             1.04   
Montfleury                                                        0.20   
Corneille                Pierre                                  -0.49   

                                                  sigma_iarkho  polylogues  \
last_name                first_name                                          
Hauteroche               Noël Lebreton, sieur de          0.41        0.88   
Boisrobert               François Le Métel de             0.61        1.44   
La Fontaine              Jean de                          1.12        0.11   
Gillet de la Tessonnerie N.                               0.81        1.26   
Ouville                  Antoine d'                       0.23        0.76   
Corneille                Thomas                          -0.22        1.17   
Scarron                  Paul                             0.75        0.61   
Boursault                Edmé                            -0.05        0.06   
Molière                                                   0.61        0.00   
Quinault                 Philippe                        -0.48        0.68   
Mareschal                André                            0.38        1.18   
Baron                                                    -0.60       -0.27   
Montfleury                                               -0.68       -0.38   
Corneille                Pierre                          -0.77       -1.12   

                                                  monologues  open_form_score  
last_name                first_name                                            
Hauteroche               Noël Lebreton, sieur de        0.64             0.81  
Boisrobert               François Le Métel de           1.67             0.67  
La Fontaine              Jean de                       -0.44             0.58  

In [48]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_above_one_play_period_one[sorted_above_one_play_period_one.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 10


In [49]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_above_one_play_period_one[sorted_above_one_play_period_one.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 4


### Less Prolific

In [50]:
sorted_only_one_play_period_one = sorted_only_one.loc[period_one.last_name.unique(), :]

In [51]:
sorted_only_one_play_period_one

num_present_characters  \
last_name                 first_name                                       
Chalussay                 Le Boulanger de                           4.96   
Guérin de Bouscal         Guyon                                     0.96   
Montfleury; Corneille     ; Thomas                                  2.56   
Beys                      Charles de                                2.02   
Robbe                     Jacques                                   1.22   
Poisson                   Raymond                                   0.16   
Lambert                                                            -0.64   
Dorimond                                                            0.42   
Palaprat                  Jean de, sieur de Bigo                    0.42   
du Ryer                   Pierre                                   -0.11   
Champmeslé                                                         -0.11   
Chevalier                                                           0.42   
L. C. D.                                                            1.22   
Tristan L'Hermite         François                                 -0.11   
Gilbert                   Gabriel                                  -0.64   
Chappuzeau                Samuel                                   -0.64   
Marcel                                                             -1.18   
Scudery                   Georges de                               -0.38   
Brosse                    Abraham                                  -0.91   
Desmarets de Saint-Sorlin Jean                                     -0.64   
L'Estoile                 Claude                                   -0.38   
Chevreau                  Urbain                                   -0.64   
Rotrou                    Jean de                                  -0.38   

                                                  mobility_coefficient  \
last_name                 first_name                                     
Chalussay                 Le Boulanger de                        -2.30   
Guérin de Bouscal         Guyon                                  -2.38   
Montfleury; Corneille     ; Thomas                               -0.73   
Beys                      Charles de                             -0.32   
Robbe                     Jacques                                -0.32   
Poisson                   Raymond                                -0.49   
Lambert                                                           0.01   
Dorimond                                                         -1.14   
Palaprat                  Jean de, sieur de Bigo                  0.34   
du Ryer                   Pierre                                 -0.73   
Champmeslé                                                        0.83   
Chevalier                                                        -0.24   
L. C. D.                                                         -1.56   
Tristan L'Hermite         François                               -1.14   
Gilbert                   Gabriel                                -0.65   
Chappuzeau                Samuel                                 -1.39   
Marcel                                                           -0.90   
Scudery                   Georges de                             -1.23   
Brosse                    Abraham                                -1.23   
Desmarets de Saint-Sorlin Jean                                   -1.14   
L'Estoile                 Claude                                 -0.16   
Chevreau                  Urbain                                 -1.47   
Rotrou                    Jean de                                -0.24   

                                                  sigma_iarkho  polylogues  \
last_name                 first_name                                         
Chalussay                 Le Boulanger de                 6.64        0.23   
Guérin de Bouscal         Guyon                           2.64  

In [52]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_only_one_play_period_one[sorted_only_one_play_period_one.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 13


In [53]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_only_one_play_period_one[sorted_only_one_play_period_one.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 10


- The ratio of open-form playwrights among the "prolific" authors (with at least 2 comedies): 2.5
- The ratio of open-form playwrights among less prolific authors (with only 1 comedy): 1.3

## Period Two: Z-Scores and Open-Form Scores

In [54]:
sorted_above_one_play_period_two = sorted_above_one.loc[period_two.last_name.unique(), :]

### Prolific

In [55]:
sorted_above_one_play_period_two

num_present_characters  \
last_name              first_name                               
Demoustier             Charles-Albert                    0.02   
Dancourt                                                -0.11   
Desforges                                                0.02   
Collin d’Harleville    Jean-François                    -0.16   
Regnard                Jean-François                     0.16   
Voltaire                                                -0.31   
Du Fresny              Charles                          -0.24   
Rousseau               Jean-Baptiste                    -0.11   
Néricault Destouches                                    -0.27   
Baron                                                    0.42   
Piron                  Alexis                           -0.78   
Dorat                  Claude-Joseph                     0.02   
Fabre d'Églantine                                       -0.38   
Boissy                 Louis de                         -0.72   
Nivelle de la Chaussée Pierre Claude                    -0.71   

                                       mobility_coefficient  sigma_iarkho  \
last_name              first_name                                           
Demoustier             Charles-Albert                  0.46          2.12   
Dancourt                                               0.70          0.26   
Desforges                                              0.48          0.58   
Collin d’Harleville    Jean-François                   0.96         -0.20   
Regnard                Jean-François                   0.27          0.23   
Voltaire                                              -0.88          0.26   
Du Fresny              Charles                         0.26         -0.08   
Rousseau               Jean-Baptiste                  -0.84          0.52   
Néricault Destouches                                   0.18          0.09   
Baron                                                  1.04         -0.60   
Piron                  Alexis                          0.79         -0.31   
Dorat                  Claude-Joseph                  -0.28          0.29   
Fabre d'Églantine                                     -0.03         -0.45   
Boissy                 Louis de                       -0.19         -0.77   
Nivelle de la Chaussée Pierre Claude                  -0.42         -0.57   

                                       polylogues  monologues  open_form_score  
last_name              first_name                                               
Demoustier             Charles-Albert        0.55       -0.03             0.62  
Dancourt                                     0.68        0.41             0.39  
Desforges                                    0.02       -0.24             0.17  
Collin d’Harleville    Jean-François         0.16        0.07             0.17  
Regnard                Jean-François         0.17       -0.12             0.14  
Voltaire                                     0.47        0.77             0.06  
Du Fresny              Charles               0.09        0.20             0.05  
Rousseau               Jean-Baptiste         0.40        0.18             0.03  
Néricault Destouches                        -0.12       -0.29            -0.08  
Baron                                       -0.27       -1.03            -0.09  
Piron                  Alexis               -0.30       -0.68            -0.26  
Dorat                  Claude-Joseph        -1.08       -0.53            -0.32  
Fabre d'Églantine                           -0.89       -0.69            -0.49  
Boissy                 Louis de             -0.83       -0.38            -0.58  
Nivelle de la Chaussée Pierre Claude        -1.04       -0.72            -0.69

In [56]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_above_one_play_period_two[sorted_above_one_play_period_two.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 8


In [57]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_above_one_play_period_two[sorted_above_one_play_period_two.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 7


### Less Prolific

In [58]:
sorted_only_one_play_period_two = sorted_only_one.loc[period_two.last_name.unique(), :]

In [59]:
sorted_only_one_play_period_two

num_present_characters  \
last_name              first_name                                            
Chauveau                                                              2.29   
Cubières-Palmézeaux    Michel de                                      3.36   
Romagnesi              Jean-Antoine                                   0.16   
Patrat                 Joseph                                         0.42   
Laya                   Jean-Louis                                     0.16   
Autreau                Jacques                                        0.96   
Dampierre de la Salle                                                 0.16   
Cailhava D'Estendoux   Jean-François                                  0.16   
Palissot de Montenoy   Charles                                       -0.38   
Chabanon               Michel Paul Guy de                             1.49   
Lantier                Étienne François de                           -0.38   
Ch***                                                                -0.64   
Bièvre                 Marquis de                                    -0.11   
Moissy                 Alexandre-Guillaume de                        -0.91   
Dupuis                 Amable-Joseph                                  0.16   
Barthe                 Nicolas-Thomas                                 0.16   
Billard-Dumonceau      Edme                                          -1.18   
de Neufchâteau         Nicolas François                              -0.64   
de Rochefort           Guillaume Dubois                              -0.64   
Campistron             Jean Galbert de                               -0.64   
Fagan                  Barthélemy-Christophe                         -0.64   
Montesson              Madame de                                      0.16   
Rochon de Chabannes    Marc-Antoine-Jacques                          -0.91   
Borel                                                                -0.11   
Dubuisson              Pierre-Ulric                                  -0.38   
Bouy; Fétiq                                                          -0.38   
Reynier                L.                                            -0.38   
Descazeaux Des Granges                                               -0.38   
La Noue                Jean-Baptiste                                 -0.64   
Gresset                Jean-Baptiste                                 -0.64   
Courtial                                                             -0.91   
Imbert                 Barthélemy                                    -0.64   
Pieyre                 Alexandre                                     -0.64   
Vigée                  Louis-Jean-Baptiste-Étienne                   -0.38   
Delon                                                                 0.69   
Bret                   Antoine                                       -1.18   
Bastide                Jean-François de                              -1.44   

                                                    mobility_coefficient  \
last_name              first_name                                          
Chauveau                                                            3.55   
Cubières-Palmézeaux    Michel de                                    0.59   
Romagnesi              Jean-Antoine                                 1.99   
Patrat                 Joseph                                       1.74   
Laya                   Jean-Louis                                  -1.31   
Autreau                Jacques                                      1.00   
Dampierre de la Salle                                              -0.73   
Cailhava D'Estendoux   Jean-François                                0.42   
Palissot de Montenoy   Charles                                     -0.16   
Chabanon               Michel Paul Guy de                           0.67   
Lantier                Étienne François de                          2.73   
Ch***                

In [60]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_only_one_play_period_two[sorted_only_one_play_period_two.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 7


In [61]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_only_one_play_period_two[sorted_only_one_play_period_two.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 30


- The ratio of open-form playwrights among the "prolific" authors (with at least 2 comedies): 1.14
- The ratio of open-form playwrights among less prolific authors (with only 1 comedy): 0.23, for every open-form playwright there are over 4 playwrights who prefer closed forms. 

## Period Three: Z-Scores and Open-Form Scores

### Playwrights Who Wrote At Least Two Five-Act Verse Comedies

In [62]:
sorted_above_one_play_period_three = sorted_above_one.loc[period_three.last_name.unique(), :]

In [63]:
sorted_above_one_play_period_three

num_present_characters  \
last_name           first_name                                  
Gosse               Étienne                              2.02   
Grille              François-Joseph                      3.22   
Merville                                                 0.29   
Delavigne           Casimir                              0.82   
Epagny              Violet d'                           -0.38   
Picard              Louis-Benoît                         0.22   
Étienne             Charles-Guillaume                   -0.11   
Bonjour             Casimir                             -0.00   
Théaulon            Emmanuel                            -0.11   
Collin d’Harleville Jean-François                       -0.16   
Riboutté            François-Louis                      -0.38   
Rey                 Charles                              0.02   
Andrieux            François                            -0.11   
Duval               Alexandre                           -0.80   
Fabre d'Églantine                                       -0.38   

                                       mobility_coefficient  sigma_iarkho  \
last_name           first_name                                              
Gosse               Étienne                            2.27          1.35   
Grille              François-Joseph                    1.00         -0.20   
Merville                                               0.42          0.81   
Delavigne           Casimir                           -0.11          1.26   
Epagny              Violet d'                          0.59          0.66   
Picard              Louis-Benoît                       0.85         -0.11   
Étienne             Charles-Guillaume                  0.92         -0.48   
Bonjour             Casimir                            1.52          0.81   
Théaulon            Emmanuel                           0.30          0.72   
Collin d’Harleville Jean-François                      0.96         -0.20   
Riboutté            François-Louis                    -0.27          0.95   
Rey                 Charles                            1.24          0.23   
Andrieux            François                           0.13          0.81   
Duval               Alexandre                         -0.12         -0.08   
Fabre d'Églantine                                     -0.03         -0.45   

                                       polylogues  monologues  open_form_score  
last_name           first_name                                                  
Gosse               Étienne                 -0.19       -0.35             1.02  
Grille              François-Joseph          0.19        0.39             0.92  
Merville                                     1.71        0.50             0.75  
Delavigne           Casimir                  0.02       -0.26             0.35  
Epagny              Violet d'                0.46        0.03             0.27  
Picard              Louis-Benoît             0.15        0.21             0.26  
Étienne             Charles-Guillaume        0.42        0.55             0.26  
Bonjour             Casimir                 -0.16       -0.87             0.26  
Théaulon            Emmanuel                 0.11       -0.18             0.17  
Collin d’Harleville Jean-François            0.16        0.07             0.17  
Riboutté            François-Louis           0.49       -0.11             0.14  
Rey                 Charles                 -0.59       -0.27             0.13  
Andrieux            François                -0.20       -0.07             0.11  
Duval               Alexandre                0.24       -0.29            -0.21  
Fabre d'Églantine                           -0.89       -0.69            -0.49

In [64]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_above_one_play_period_three[sorted_above_one_play_period_three.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 13


In [65]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_above_one_play_period_three[sorted_above_one_play_period_three.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 2


### Less Prolific

In [66]:
sorted_only_one_play_period_three = sorted_only_one.loc[period_three.last_name.unique(), :]

In [67]:
sorted_only_one_play_period_three

num_present_characters  \
last_name           first_name                                     
Girardin            Delphine de                             3.89   
Viennet             Jean-Pons-Guillaume                    -0.11   
Roux                                                        0.42   
Dalban              Pierre-Jean-Baptiste                    0.42   
Bernay              Camille                                 1.49   
de Planard          Eugène                                  0.69   
Princeteau          Th.                                     0.42   
Poya                Pierre                                  0.96   
La Ville de Mirmon  Jean de                                -0.38   
Mery                Joseph                                  0.42   
Charlemagne         Jean-Armand                            -0.11   
Triquet             C                                       0.42   
Drap-Arnaud                                                -0.11   
Mennechet           Édouard                                -0.38   
La Ville de Mirmont Jean de                                -0.38   
Leroy               Onésime                                -0.38   
Augier              Émile                                  -1.44   
Dalban              Pierre Jean Baptiste                   -0.64   
Roucher             J.F.                                   -0.91   
Michaud             L. G.                                  -0.11   
Capuran             Louis-Benoit                           -0.64   
Ancien militaire                                           -0.64   

                                          mobility_coefficient  sigma_iarkho  \
last_name           first_name                                                 
Girardin            Delphine de                           0.09          2.58   
Viennet             Jean-Pons-Guillaume                   1.41          1.84   
Roux                                                      0.01          1.12   
Dalban              Pierre-Jean-Baptiste                  0.42          1.01   
Bernay              Camille                               1.57          0.15   
de Planard          Eugène                                0.42          0.41   
Princeteau          Th.                                  -0.16          1.46   
Poya                Pierre                               -0.07         -0.17   
La Ville de Mirmon  Jean de                               0.09          0.29   
Mery                Joseph                                0.26         -0.08   
Charlemagne         Jean-Armand                           1.99         -0.20   
Triquet             C                                     0.50          0.09   
Drap-Arnaud                                               1.24         -0.48   
Mennechet           Édouard                               0.34         -0.22   
La Ville de Mirmont Jean de                               0.75         -0.05   
Leroy               Onésime                               1.16          0.15   
Augier              Émile                                -0.32         -0.34   
Dalban              Pierre Jean Baptiste                  0.09         -0.02   
Roucher             J.F.                                  0.67         -0.17   
Michaud             L. G.                                -1.06         -1.08   
Capuran             Louis-Benoit                         -0.49         -0.88   
Ancien militaire                                          0.34         -0.51   

                                          polylogues  monologues  \
last_name           first_name                                     
Girardin            Delphine de                -0.19       -0.67   
Viennet             Jean-Pons-Guillaume         0.97        0.34   
Roux                                            1.86        0.43   
Dalban              Pierre-Jean-Baptiste        0.50        0.59   
Bernay              Camille                     0.02       -0.44   
de Planard  

In [68]:
print('Number of playwrigthts writing comedies with the open form:', 
sorted_only_one_play_period_three[sorted_only_one_play_period_three.open_form_score >= 0].shape[0])

Number of playwrigthts writing comedies with the open form: 12


In [69]:
print('Number of playwrigthts writing comedies with the closed form:', 
sorted_only_one_play_period_three[sorted_only_one_play_period_three.open_form_score < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 10


- The ratio of open-form playwrights among the "prolific" authors (with at least 2 comedies): 6.5
- The ratio of open-form playwrights among less prolific authors (with only 1 comedy): 1.2.

## Part 3. The Evolution of Jean-François Collin d’Harleville (1755-805)

In [70]:
print('Number dramatic characters in period two:',  
      round(period_two[period_two.last_name=='Collin d’Harleville']['num_present_characters'].mean(), 2))
print('Number dramatic characters in period three:', 
      period_three[period_three.last_name=='Collin d’Harleville']['num_present_characters'].mean())

Number dramatic characters in period two: 10.67
Number dramatic characters in period three: 11.0


In [71]:
print('Mobility coefficient in period two:',
      round(period_two[period_two.last_name=='Collin d’Harleville']['mobility_coefficient'].mean(), 2))
print('Mobility coefficient in period three:', 
      period_three[period_three.last_name=='Collin d’Harleville']['mobility_coefficient'].mean())

Mobility coefficient in period two: 60.33
Mobility coefficient in period three: 58.5


In [72]:
print('Standard range of the number of speaking characters in period two:', 
      period_two[period_two.last_name=='Collin d’Harleville'].pipe(speech_distribution_by_period).pipe(sigma_iarkho))
print('Standard range of the number of speaking characters in period three:',
       period_three[period_three.last_name=='Collin d’Harleville'].pipe(speech_distribution_by_period).pipe(sigma_iarkho))

Standard range of the number of speaking characters in period two: 1.07
Standard range of the number of speaking characters in period three: 1.24


In [73]:
print('The percentage of polylogues in period two:', 
     round(period_two[period_two.last_name=='Collin d’Harleville']['percentage_polylogues'].mean(), 2))
print('The percentage of polylogues in period three:',
     round(period_three[period_three.last_name=='Collin d’Harleville']['percentage_polylogues'].mean(), 2))

The percentage of polylogues in period two: 36.26
The percentage of polylogues in period three: 43.22


In [74]:
print('The percentage of monologues in period two:',
     round(period_two[period_two.last_name=='Collin d’Harleville']['percentage_monologues'].mean(), 2))
print('The percentage of monologues in period three:',
     round(period_three[period_three.last_name=='Collin d’Harleville']['percentage_monologues'].mean(), 2))

The percentage of monologues in period two: 18.99
The percentage of monologues in period three: 16.18


Collin d’Harleville's comedies spanned two periods: two and three. He tended to be amongs moderately experimental playwrights (his open-form score is 0.17).  The evolution of his comedies is fascinating as it echoes broader literary trends in reguards to all features except for mobility coefficient. The number of dramatic characters increases from period two (10.64) to period three (11.96). The mean number of dramatic characters in Collin d’Harleville's comedies increases from 10.67 in period two to 11 in period three. Standard range of the number of speaking characters increases from 1.12 in period two to 1.33 in period three. The standard range in Collin d’Harleville's comedies increases from 1.07 in period two to 1.24 in period three. The percentage of polylogues increases from 31.37 in period two to 37.69 in period three. The percentage of polylogues increases in Collin d’Harleville's comedies increases from 36.26 in period two to 43.22 in period three. The percentage of monologues decreases from 22.08 in period two to 20.35 in period three. The percentage of monologues in Collin d’Harleville's comedies decreases from 18.99 to 16.18. The mobility coefficient increases from 49.56 in period two to 54.31 in period three. Collin d’Harleville's comedies, however, tend to have higher mobility coefficient in period two (60.33) and slightly lower in period three (58.5), but in both cases, he is above the mean moblity coefficient of the period.

## Conclusions:

1. For each feature and each time period, we identified the comedies that represent the maxim and the minimum as well as such comedies that are the closest to the mean.
2. We discovered that that there are at least nine comedies without monologues (all in period one).
3. Three authors (Chalussay, Rousseau, and Gosse) wrote the comedies that represent the maximum for their period based on at least two features.
4. Two authors (Pierre Corneille and Casimir Bonjour) wrote comedies that represent their period minimum based on at least two features.
5. We analyzed Jean-François Collin d’Harleville's style as it evolved over two periods and echoed wider literary trends. 
6. We introduced **"open-form" scores** that capture how experimental each playwright in the context of the entire history of the French five-act comedy in verse. In aggregate, the number of more experimental and more conservative playwrights was balanced: 62 to 61. However, the situation became more complex as we subdivided the playwrights into more prolific and less prolific and analyzed them by periods.
    - **Period one**. In the prolific subgroup, the number of experimental playwrights is 2.5 as high as the number of classical playwrights (10 vs. 4). In the less prolific subgroup, the number of experimental playwrights is 1.3 as high as the number of classical playwrights (13 vs. 10).
    - **Period two**. In the prolific subgroup, the number of experimental playwrights is 1.14 as high as the number of classical playwrights (8 vs. 7). In the less prolific subgroup, the opposite is true: the number of conservative playwrights is over 4 times as high as the number of experimental playwrights (7 vs. 30).
    - **Period three**. In the prolific subgroup, the number of experimental playwrights is 6.5 as high as the number of conservative playwrights (13 vs. 2). In the less prolific subgroup, the number of experimental playwrights is 1.2 as high as the number of conservative playwrights (12 vs. 10). 
    Overall, the prolific playwrights are the minority: there are twice as many playwrights who wrote just one five-act verse comedy (41 vs 82 less prolific). The prolific playwrights tend to be more experimental (30 experimental authors vs. 11 conservative) than the playwrights in the less prolific group (32 experimental vs. 52 conservative), which is true for every period examined.  

## Appendix

In [75]:
results_means = playwrights_place(sorted_comedies, with_z_score=False).loc[results.index, :]
results_means['open_form_score'] = results.open_form_score.tolist()

### Raw Numbers and Open-Form Scores

In [76]:
pd.set_option('display.max_rows', 200)
display(results_means)

num_present_characters  \
last_name                 first_name                                            
Chalussay                 Le Boulanger de                               30.00   
Guérin de Bouscal         Guyon                                         15.00   
Girardin                  Delphine de                                   26.00   
Gosse                     Étienne                                       19.00   
Montfleury; Corneille     ; Thomas                                      21.00   
Beys                      Charles de                                    19.00   
Grille                    François-Joseph                               23.50   
Viennet                   Jean-Pons-Guillaume                           11.00   
Robbe                     Jacques                                       16.00   
Hauteroche                Noël Lebreton, sieur de                       13.75   
Roux                                                                    13.00   
Merville                                                                12.50   
Boisrobert                François Le Métel de                          13.17   
Poisson                   Raymond                                       12.00   
Chauveau                                                                20.00   
Cubières-Palmézeaux       Michel de                                     24.00   
Demoustier                Charles-Albert                                11.50   
Dalban                    Pierre-Jean-Baptiste                          13.00   
La Fontaine               Jean de                                       19.00   
Bernay                    Camille                                       17.00   
de Planard                Eugène                                        14.00   
Romagnesi                 Jean-Antoine                                  12.00   
Gillet de la Tessonnerie  N.                                            13.67   
Ouville                   Antoine d'                                    10.33   
Lambert                                                                  9.00   
Dorimond                                                                13.00   
Dancourt                                                                11.00   
Corneille                 Thomas                                        11.25   
Palaprat                  Jean de, sieur de Bigo                        13.00   
Delavigne                 Casimir                                       14.50   
Patrat                    Joseph                                        13.00   
du Ryer                   Pierre                                        11.00   
Epagny                    Violet d'                                     10.00   
Picard                    Louis-Benoît                                  12.25   
Étienne                   Charles-Guillaume                             11.00   
Bonjour                   Casimir                                       11.40   
Boursault                 Edmé                                          16.00   
Scarron                   Paul                                          11.50   
Princeteau                Th.                                           13.00   
Champmeslé                                                              11.00   
Théaulon                  Emmanuel                                      11.00   
Collin d’Harleville       Jean-François                                 10.80   
Desforges                                                               11.50   
Laya                      Jean-Louis                                    12.00   
Autreau                   Jacques                                       15.00   
Poya                      Pierre                                        15.00   
Riboutté                  François-Louis                                10.00   
Regnard                   Jean-François                                 12.00   
La Vi

### Z-Scores and Open-Form Scores

In [77]:
pd.set_option('display.max_rows', 200)
results

num_present_characters  \
last_name                 first_name                                            
Chalussay                 Le Boulanger de                                4.96   
Guérin de Bouscal         Guyon                                          0.96   
Girardin                  Delphine de                                    3.89   
Gosse                     Étienne                                        2.02   
Montfleury; Corneille     ; Thomas                                       2.56   
Beys                      Charles de                                     2.02   
Grille                    François-Joseph                                3.22   
Viennet                   Jean-Pons-Guillaume                           -0.11   
Robbe                     Jacques                                        1.22   
Hauteroche                Noël Lebreton, sieur de                        0.62   
Roux                                                                     0.42   
Merville                                                                 0.29   
Boisrobert                François Le Métel de                           0.47   
Poisson                   Raymond                                        0.16   
Chauveau                                                                 2.29   
Cubières-Palmézeaux       Michel de                                      3.36   
Demoustier                Charles-Albert                                 0.02   
Dalban                    Pierre-Jean-Baptiste                           0.42   
La Fontaine               Jean de                                        2.02   
Bernay                    Camille                                        1.49   
de Planard                Eugène                                         0.69   
Romagnesi                 Jean-Antoine                                   0.16   
Gillet de la Tessonnerie  N.                                             0.60   
Ouville                   Antoine d'                                    -0.29   
Lambert                                                                 -0.64   
Dorimond                                                                 0.42   
Dancourt                                                                -0.11   
Corneille                 Thomas                                        -0.04   
Palaprat                  Jean de, sieur de Bigo                         0.42   
Delavigne                 Casimir                                        0.82   
Patrat                    Joseph                                         0.42   
du Ryer                   Pierre                                        -0.11   
Epagny                    Violet d'                                     -0.38   
Picard                    Louis-Benoît                                   0.22   
Étienne                   Charles-Guillaume                             -0.11   
Bonjour                   Casimir                                       -0.00   
Boursault                 Edmé                                           1.22   
Scarron                   Paul                                           0.02   
Princeteau                Th.                                            0.42   
Champmeslé                                                              -0.11   
Théaulon                  Emmanuel                                      -0.11   
Collin d’Harleville       Jean-François                                 -0.16   
Desforges                                                                0.02   
Laya                      Jean-Louis                                     0.16   
Autreau                   Jacques                                        0.96   
Poya                      Pierre                                         0.96   
Riboutté                  François-Louis                                -0.38   
Regnard                   Jean-François                                  0.16   
La Vi

In [78]:
print('Number of playwrigthts writing comedies with the open form:', 
results[results.open_form_score >=0].shape[0])

Number of playwrigthts writing comedies with the open form: 62


In [79]:
print('Number of playwrigthts writing comedies with the closed form:', 
results[results.open_form_score  < 0].shape[0])

Number of playwrigthts writing comedies with the closed form: 61
